RavenDB-14325 Tombstone cleaner should take into account backups and ETLs cluster-wide

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-14325 Tombstone cleaner should take into account backups

We added the database change vector to the backup status.
This allow us to verify cluster-wide which etag of the local node was backed up.

karmeli87 committed:
RavenDB-14325 Tombstone cleaner should take into account ETLs

The cleaner is subscribed globally for an etl loader (not for etl process).
When have to ensure that all of the etls (cluster-wide) loaded all tombstones up-to the etag element of the current node.

karmeli87 committed:
RavenDB-14325 add tests

ayende commented on src/Raven.Server/Documents/DocumentDatabase.cs:
```diff
@@ -867,6 +867,15 @@ public long ReadLastEtag()
+            using (DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext documentsContext))
+            using (var tx = documentsContext.OpenReadTransaction())
+            {
+                return (DocumentsStorage.ReadLastEtag(tx.InnerTransaction), DocumentsStorage.GetDatabaseChangeVector(tx.InnerTransaction));
```
Is there a match between the etag and the change vector?
_should_ there be?

karmeli87 replied:
I think there should be a match.
But, does it really matter?

ayende commented on src/Raven.Server/Documents/ETL/EtlProcess.cs:
```diff
@@ -814,7 +810,7 @@ protected void EnsureThreadAllocationStats()
         }
 
-        public override Dictionary<string, long> GetLastProcessedTombstonesPerCollection()
+        /*public override Dictionary<string, long> GetLastProcessedTombstonesPerCollection()
```
why leave commented code?

karmeli87 replied:
removed

ppekrol commented:
@grisha-kotler please check Backup part
@arekpalinski please check ETL part

karmeli87 committed:
RavenDB-14325 localServer shouldn't part of the Servers list

arekpalinski commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -19,7 +22,7 @@
 namespace Raven.Server.Documents.ETL
 {
-    public class EtlLoader : IDisposable
+    public class EtlLoader : IDisposable, ITombstoneAware
```
Why do we need to make the loaded aware of tomstones? Why was it not enough that each process implemented `ITombstoneAware`?

karmeli87 replied:
Because even node that isn't responsible for the ETL should take it into account.
It can't purge the tombstone until the responsible  node sent it successfully.

arekpalinski replied:
Oh, I see it now. Thanks

arekpalinski commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -492,5 +491,60 @@ public void HandleDatabaseValueChanged(DatabaseRecord record)
+            foreach (var transform in config.Transforms)
+            {
+                var state = EtlProcess.GetProcessState(_database, config.Name, transform.Name);
+                var etag = ChangeVectorUtils.GetEtagById(state.ChangeVector, _database.DbBase64Id);
```
Why you no longer use `EtlProcessState.GetLastProcessedEtagForNode` ?

karmeli87 replied:
This dictionary is updated on the responsible node.
How the other nodes should know if it is okay to delete their tombstones?

arekpalinski replied:
Makes sense.

grisha-kotler commented on src/Raven.Server/Documents/PeriodicBackup/PeriodicBackupRunner.cs:
```diff
@@ -634,13 +635,37 @@ private PeriodicBackupStatus GetBackupStatusFromCluster(long taskId)
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                var record = _serverStore.Cluster.ReadRawDatabaseRecord(context, _database.Name);
```
When we have a backup with `Full Backup/Snapshot` without an incremental, we don't need to keep the tombstones at all.

karmeli87 replied:
I think we do need to keep the tombstones... consider we have an ETL or external replication task that didn't sent the tombstones yet.

grisha-kotler replied:
We don't backup tombstones during a full backup

karmeli87 replied:
done

karmeli87 committed:
RavenDB-14325 don't account _only_ full backup configuration

if the backup is always full (no incremental configuration), we don't need to take into account the tombstones, since we never back them up.
